### PR TITLE
Support trees in lax parallel operations.

### PR DESCRIPTION
**Pull request by tomhennigan**
It is relatively common to apply collective operations to trees. For example in sync distributed training it is typical to sum all gradients across replicas `grads = jax.tree_map(partial(lax.psum, axis_name='i'), grads)`. We can make this a little more convenient by making lax parallel ops support trees directly: `grads = lax.psum(grads, 'i')`.

There is room for improvement in this change. We should in some (all?) cases just pass a tuple of values to XLA (rather than bind the primivive `n` times bind once with a tuple of `n` values) however this produced strange values when combined with `pmap` and a fix was not obvious. This is something we can follow up on without users having to change their code.

### CODE REVIEW

**Comment by skye** [comment]
Nice! Do you mind adding some test cases? I think one per primitive should be fine, if that.

**Comment by tomhennigan** [comment]
Sorry for the delay, very happy to add tests just didn't get to it this week and will be OOO for a week. Will sort this when I get back 😄

**Comment by j-towns** [comment]
+1 for this @tomhennigan

**Comment by tomhennigan** [comment]
Sorry for the delay, have added tests covering `pmax`, `pmin`, `psum`, `ppermute` and `all_to_all`.

**Comment by j-towns** [comment]
Might also be worth altering the docstrings to reflect these functions’ new flexibility.

**Comment by tomhennigan** [comment]
> Might also be worth altering the docstrings to reflect these functions’ new flexibility.

Done.

It would be great to add some type annotations here, but JAX haven't added them elsewhere (probably because until recently they supported Python 2), I'll add them in another PR. I think the following would make sense (might need to chuck an `Any` in there for custom tree types):

```python
ArrayTree = Union[jnp.ndarray, Tuple["ArrayTree", ...], List["ArrayTree"], Mapping[Any, "ArrayTree"]]

T = TypeVar('T', ATree)
def psum(x: T, axis_name) -> T:
  ...
```

**Comment by j-towns** [comment]
Ping @skye

**Comment by hawkinsp** [comment]
@tomhennigan FYI: I did start adding a few type annotations to JAX once we dropped Python 2. So there's nothing stopping us from adding more. I only did the minimimum necessary to make PyType not complain, nothing more.

**Comment by tomhennigan** [comment]
Awesome! I'll still add them in a follow up to avoid blocking this change for longer 😄.

**Review by skye — APPROVED** [review]
Sorry for the delay! Looks great, thanks!
